Support animations in Model element with model process

achan00 opened the pull request:
#### 845d49cbfca7509085faa9620bd1b942a6ff709e
<pre>
Support animations in Model element with model process
<a href="https://bugs.webkit.org/show_bug.cgi?id=280312">https://bugs.webkit.org/show_bug.cgi?id=280312</a>
<a href="https://rdar.apple.com/114258826">rdar://114258826</a>

Reviewed by Mike Wyrzykowski.

Initial implementation of animations API that is based on the latest &lt;model&gt;
explainer where animation properties follow more closely to similar ones in
the media element. These API are implemented for the model process only.

This initial implementation is based on a single animation track (first animation
found in the model). Properties supported are: autoplay, loop, playbackRate,
duration, paused, currentTime. Methods supported are play() and pause().

* LayoutTests/model-element/model-element-animations-current-time-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-current-time.html: Added.
* LayoutTests/model-element/model-element-animations-initial-values-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-initial-values.html: Added.
* LayoutTests/model-element/model-element-animations-loop-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-loop.html: Added.
* LayoutTests/model-element/model-element-animations-playback-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-playback.html: Added.
* LayoutTests/model-element/model-element-animations-replace-sources-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-replace-sources.html: Added.
* LayoutTests/model-element/resources/model-utils.js:
(sleepForSeconds):
* LayoutTests/model-element/resources/stopwatch-30s.usdz: Added.
* LayoutTests/model-element/resources/stopwatch-60s.usdz: Added.
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::createModelPlayer):
Set up the model player with the initial autoplay, loop, and playbackRate values.
(WebCore::HTMLModelElement::attributeChanged):
(WebCore::HTMLModelElement::setPlaybackRate):
(WebCore::HTMLModelElement::duration const):
(WebCore::HTMLModelElement::paused const):
(WebCore::HTMLModelElement::play):
(WebCore::HTMLModelElement::pause):
(WebCore::HTMLModelElement::setPaused):
(WebCore::HTMLModelElement::autoplay const):
(WebCore::HTMLModelElement::updateAutoplay):
(WebCore::HTMLModelElement::loop const):
(WebCore::HTMLModelElement::updateLoop):
(WebCore::HTMLModelElement::currentTime const):
(WebCore::HTMLModelElement::setCurrentTime):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::setAutoplay):
(WebCore::ModelPlayer::setLoop):
(WebCore::ModelPlayer::setPlaybackRate):
(WebCore::ModelPlayer::duration const):
(WebCore::ModelPlayer::paused const):
(WebCore::ModelPlayer::setPaused):
(WebCore::ModelPlayer::currentTime const):
(WebCore::ModelPlayer::setCurrentTime):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
Introduce WKModelProcessModelPlayerProxyObjCAdapter which implements WKSRKEntityDelegate
for getting callbacks when the animation playback state changes.
(-[WKModelProcessModelPlayerProxyObjCAdapter initWithModelProcessModelPlayerProxy:]):
(-[WKModelProcessModelPlayerProxyObjCAdapter entityAnimationPlaybackStateDidUpdate:]):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::startAnimating):
Pass in the autoplay state when setting up the animation. Set up the entity
with the loop and playbackRate values.
(WebKit::ModelProcessModelPlayerProxy::animationPlaybackStateDidUpdate):
Send the latest animation playback state back to ModelProcessModelPlayer in the WP.
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::setAutoplay):
(WebKit::ModelProcessModelPlayerProxy::setLoop):
(WebKit::ModelProcessModelPlayerProxy::setPlaybackRate):
(WebKit::ModelProcessModelPlayerProxy::duration const):
(WebKit::ModelProcessModelPlayerProxy::paused const):
(WebKit::ModelProcessModelPlayerProxy::setPaused):
(WebKit::ModelProcessModelPlayerProxy::currentTime const):
(WebKit::ModelProcessModelPlayerProxy::setCurrentTime):
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.delegate):
(WKSRKEntity.animationPlaybackController):
(WKSRKEntity.animationFinishedSubscription):
(WKSRKEntity._duration):
(WKSRKEntity._playbackRate):
(WKSRKEntity.duration):
(WKSRKEntity.loop):
(WKSRKEntity.playbackRate):
(WKSRKEntity.paused):
(WKSRKEntity.currentTime):
(WKSRKEntity.setUpAnimation(with:)):
If an animation track is found in the entity, call playAnimation with startsPaused
based on the current autoplay state. That returns an animation playback controller
that we can use to play/pause this current animation, set its playback rate, and
query its current time. Subscribe to the PlaybackCompleted animation event, and on
animation completion, we can set up the next animation based on the current
loop and playbackRate values.
(WKSRKEntity.animationPlaybackStateDidUpdate):
(WKSRKEntity.startAnimating): Deleted.
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::sendWithAsyncReply):
(WebKit::ModelProcessModelPlayer::didUpdateAnimationPlaybackState):
(WebKit::ModelProcessModelPlayer::setAutoplay):
(WebKit::ModelProcessModelPlayer::setLoop):
(WebKit::ModelProcessModelPlayer::setPlaybackRate):
(WebKit::ModelProcessModelPlayer::duration const):
(WebKit::ModelProcessModelPlayer::paused const):
(WebKit::ModelProcessModelPlayer::setPaused):
(WebKit::ModelProcessModelPlayer::currentTime const):
We make our best effort to estimate the current time based on the last reported
animation time, playback rate, and duration from the model process, and the clock
timestamp of that last update. If the web content just manually set the
current time and it&apos;s still pending being handled by the model process, return
that previously set time.
(WebKit::ModelProcessModelPlayer::setCurrentTime):
Clamp the set time to be between 0 and the animation duration. Store the time being
set to m_pendingCurrentTime, and clear it in the completion handler.

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:

Canonical link: <a href="https://commits.webkit.org/284359@main">https://commits.webkit.org/284359@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de72879b07e1a6802ae896070a578b56e1b32408

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68927 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48328 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21600 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73007 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20079 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71045 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54901 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13345 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71993 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44129 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59512 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35371 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40794 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16942 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18460 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62743 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17287 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74714 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12910 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16533 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62535 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59595 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62437 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15343 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10411 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4021 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44130 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45205 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44946 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -596,6 +609,87 @@ void HTMLModelElement::setCamera(HTMLModelElementCamera camera, DOMPromiseDeferr
 // MARK: - Animations support.
 
+#if ENABLE(MODEL_PROCESS)
+void HTMLModelElement::setPlaybackRate(double playbackRate)
+{
+    if (m_playbackRate == playbackRate)
+        return;
+
+    m_playbackRate = playbackRate;
+
+    if (m_modelPlayer)
+        m_modelPlayer->setPlaybackRate(playbackRate, [&](double) { });
+}
```
In the Web Animations spec, there is the distinct notion of the `playbackRate` and an `effectivePlaybackRate` where the former returns the value that is in use at the moment it is requested and thus might differ from what was last set via the API, and the latter whatever was last set via the API, even if it it yet to be committed.

achan00 replied:
With the current code changes, the requested playback rate and the effective playback rate are handled within `ModelProcessModelPlayer`. I'm not sure if that logic should be moved to `HTMLModelElement` instead or if each `ModelPlayer` implementation should decide how they want to implement the animation details.

achan00 replied:
@graouts Do you think this is OK for now until there is clearer text in the spec on handling playback rate?

graouts replied:
Yes, this is by no mean blocking feedback, just something to consider for asynchronous properties such as `playbackRate`.

achan00 committed:
Support animations in Model element with model process
https://bugs.webkit.org/show_bug.cgi?id=280312
rdar://114258826

Reviewed by Mike Wyrzykowski.

Initial implementation of animations API that is based on the latest <model>
explainer where animation properties follow more closely to similar ones in
the media element. These API are implemented for the model process only.

This initial implementation is based on a single animation track (first animation
found in the model). Properties supported are: autoplay, loop, playbackRate,
duration, paused, currentTime. Methods supported are play() and pause().

* LayoutTests/model-element/model-element-animations-current-time-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-current-time.html: Added.
* LayoutTests/model-element/model-element-animations-initial-values-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-initial-values.html: Added.
* LayoutTests/model-element/model-element-animations-loop-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-loop.html: Added.
* LayoutTests/model-element/model-element-animations-playback-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-playback.html: Added.
* LayoutTests/model-element/model-element-animations-replace-sources-expected.txt: Added.
* LayoutTests/model-element/model-element-animations-replace-sources.html: Added.
* LayoutTests/model-element/resources/model-utils.js:
(sleepForSeconds):
* LayoutTests/model-element/resources/stopwatch-30s.usdz: Added.
* LayoutTests/model-element/resources/stopwatch-60s.usdz: Added.
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::createModelPlayer):
Set up the model player with the initial autoplay, loop, and playbackRate values.
(WebCore::HTMLModelElement::attributeChanged):
(WebCore::HTMLModelElement::setPlaybackRate):
(WebCore::HTMLModelElement::duration const):
(WebCore::HTMLModelElement::paused const):
(WebCore::HTMLModelElement::play):
(WebCore::HTMLModelElement::pause):
(WebCore::HTMLModelElement::setPaused):
(WebCore::HTMLModelElement::autoplay const):
(WebCore::HTMLModelElement::updateAutoplay):
(WebCore::HTMLModelElement::loop const):
(WebCore::HTMLModelElement::updateLoop):
(WebCore::HTMLModelElement::currentTime const):
(WebCore::HTMLModelElement::setCurrentTime):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::setAutoplay):
(WebCore::ModelPlayer::setLoop):
(WebCore::ModelPlayer::setPlaybackRate):
(WebCore::ModelPlayer::duration const):
(WebCore::ModelPlayer::paused const):
(WebCore::ModelPlayer::setPaused):
(WebCore::ModelPlayer::currentTime const):
(WebCore::ModelPlayer::setCurrentTime):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
Introduce WKModelProcessModelPlayerProxyObjCAdapter which implements WKSRKEntityDelegate
for getting callbacks when the animation playback state changes.
(-[WKModelProcessModelPlayerProxyObjCAdapter initWithModelProcessModelPlayerProxy:]):
(-[WKModelProcessModelPlayerProxyObjCAdapter entityAnimationPlaybackStateDidUpdate:]):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::startAnimating):
Pass in the autoplay state when setting up the animation. Set up the entity
with the loop and playbackRate values.
(WebKit::ModelProcessModelPlayerProxy::animationPlaybackStateDidUpdate):
Send the latest animation playback state back to ModelProcessModelPlayer in the WP.
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::setAutoplay):
(WebKit::ModelProcessModelPlayerProxy::setLoop):
(WebKit::ModelProcessModelPlayerProxy::setPlaybackRate):
(WebKit::ModelProcessModelPlayerProxy::duration const):
(WebKit::ModelProcessModelPlayerProxy::paused const):
(WebKit::ModelProcessModelPlayerProxy::setPaused):
(WebKit::ModelProcessModelPlayerProxy::currentTime const):
(WebKit::ModelProcessModelPlayerProxy::setCurrentTime):
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.delegate):
(WKSRKEntity.animationPlaybackController):
(WKSRKEntity.animationFinishedSubscription):
(WKSRKEntity._duration):
(WKSRKEntity._playbackRate):
(WKSRKEntity.duration):
(WKSRKEntity.loop):
(WKSRKEntity.playbackRate):
(WKSRKEntity.paused):
(WKSRKEntity.currentTime):
(WKSRKEntity.setUpAnimation(with:)):
If an animation track is found in the entity, call playAnimation with startsPaused
based on the current autoplay state. That returns an animation playback controller
that we can use to play/pause this current animation, set its playback rate, and
query its current time. Subscribe to the PlaybackCompleted animation event, and on
animation completion, we can set up the next animation based on the current
loop and playbackRate values.
(WKSRKEntity.animationPlaybackStateDidUpdate):
(WKSRKEntity.startAnimating): Deleted.
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::sendWithAsyncReply):
(WebKit::ModelProcessModelPlayer::didUpdateAnimationPlaybackState):
(WebKit::ModelProcessModelPlayer::setAutoplay):
(WebKit::ModelProcessModelPlayer::setLoop):
(WebKit::ModelProcessModelPlayer::setPlaybackRate):
(WebKit::ModelProcessModelPlayer::duration const):
(WebKit::ModelProcessModelPlayer::paused const):
(WebKit::ModelProcessModelPlayer::setPaused):
(WebKit::ModelProcessModelPlayer::currentTime const):
We make our best effort to estimate the current time based on the last reported
animation time, playback rate, and duration from the model process, and the clock
timestamp of that last update. If the web content just manually set the
current time and it's still pending being handled by the model process, return
that previously set time.
(WebKit::ModelProcessModelPlayer::setCurrentTime):
Clamp the set time to be between 0 and the animation duration. Store the time being
set to m_pendingCurrentTime, and clear it in the completion handler.

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:

Canonical link: https://commits.webkit.org/284359@main
